lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp

```diff
@@ -282,35 +387,92 @@ SILCombiner::visitUncheckedAddrCastInst(UncheckedAddrCastInst *UADCI) {
+  //
+  // NOTE: For owned values, we only perform this optimization if we can
+  // guarantee that we can eliminate the initial unchecked_ref_cast.
+  if (auto *otherURCI = dyn_cast<UncheckedRefCastInst>(urci->getOperand())) {
+    SILValue otherURCIOp = otherURCI->getOperand();
+    if (otherURCIOp.getOwnershipKind() != OwnershipKind::Owned) {
+      return Builder.createUncheckedRefCast(urci->getLoc(), otherURCIOp,
+                                            urci->getType());
+    }
+    SingleBlockOwnedForwardingInstFolder folder(*this, urci);
+    if (folder.add(otherURCI)) {
+      auto *newValue = Builder.createUncheckedRefCast(
+          urci->getLoc(), otherURCIOp, urci->getType());
+      return std::move(folder).optimizeWithReplacement(newValue);
+    }
+  }
 
-  // (unchecked-ref-cast (unchecked-ref-cast x X->Y) Y->Z)
-  //   ->
-  // (unchecked-ref-cast x X->Z)
-  if (auto *OtherURCI = dyn_cast<UncheckedRefCastInst>(URCI->getOperand()))
-    return Builder.createUncheckedRefCast(URCI->getLoc(),
-                                          OtherURCI->getOperand(),
-                                          URCI->getType());
-
-  // (unchecked_ref_cast (upcast x X->Y) Y->Z) -> (unchecked_ref_cast x X->Z)
-  if (auto *UI = dyn_cast<UpcastInst>(URCI->getOperand()))
-    return Builder.createUncheckedRefCast(URCI->getLoc(),
-                                              UI->getOperand(),
-                                              URCI->getType());
-
-  if (URCI->getType() != URCI->getOperand()->getType() &&
-      URCI->getType().isExactSuperclassOf(URCI->getOperand()->getType()))
-    return Builder.createUpcast(URCI->getLoc(), URCI->getOperand(),
-                                URCI->getType());
-
-  // (unchecked_ref_cast (open_existential_ref (init_existential_ref X))) ->
-  // (unchecked_ref_cast X)
-  if (auto *OER = dyn_cast<OpenExistentialRefInst>(URCI->getOperand()))
-    if (auto *IER = dyn_cast<InitExistentialRefInst>(OER->getOperand()))
-      return Builder.createUncheckedRefCast(URCI->getLoc(), IER->getOperand(),
-                                            URCI->getType());
+  // %0 = upcast %x : $X->Y
+  // %1 = unchecked_ref_cast %0 : $Y->Z
+  //
+  // ->
+  //
+  // %0 = upcast %x : $X->Y
+  // %1 = unchecked_ref_cast %x : $X->Z
+  //
+  // NOTE: For owned values, we only perform this optimization if we can
+  // guarantee that we can eliminate the upcast.
+  if (auto *ui = dyn_cast<UpcastInst>(urci->getOperand())) {
+    SILValue uiOp = ui->getOperand();
+
+    if (uiOp.getOwnershipKind() != OwnershipKind::Owned) {
+      return Builder.createUncheckedRefCast(urci->getLoc(), uiOp,
+                                            urci->getType());
+    }
+
+    SingleBlockOwnedForwardingInstFolder folder(*this, urci);
+    if (folder.add(ui)) {
+      auto *newValue =
+          Builder.createUncheckedRefCast(urci->getLoc(), uiOp, urci->getType());
+      return std::move(folder).optimizeWithReplacement(newValue);
```

Comment:
Why are you creating a new instruction here and using `optimizeWithReplacement`?

Reply:
I am matching what the code did previously.